r/src/arrow_types.h

```diff
@@ -0,0 +1,150 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <RcppCommon.h>
+
+#undef Free
+#include <arrow/api.h>
+#include <arrow/type.h>
+
+#define R_ERROR_NOT_OK(s) do { if(!s.ok()) Rcpp::stop(s.ToString()); } while (0);
+
+template <typename T>
+class static_ptr {
+public:
+  using element_type = T;
+
+  static_ptr(T* ptr_) : ptr(ptr_){}
+  inline T& operator*() const {
+    return *ptr;
+  }
+  inline T* operator->() const {
+    return ptr;
+  }
+
+private:
+  T* ptr;
+};
+
+namespace Rcpp{
+namespace traits{
+
+struct wrap_type_shared_ptr_tag{};
+struct wrap_type_static_ptr_tag{};
+
+template <typename T>
+struct wrap_type_traits<std::shared_ptr<T>>{
+  using wrap_category = wrap_type_shared_ptr_tag;
+};
+
+template <typename T>
+struct wrap_type_traits<static_ptr<T>>{
+  using wrap_category = wrap_type_static_ptr_tag;
+};
+
+template <typename T>
+class Exporter<std::shared_ptr<T>>;
+
+template <typename T>
+class Exporter<static_ptr<T>>;
+
+}
+namespace internal{
+
+template <typename T>
+inline SEXP wrap_dispatch(const T& x, Rcpp::traits::wrap_type_shared_ptr_tag) ;
+
+template <typename T>
+inline SEXP wrap_dispatch(const T& x, Rcpp::traits::wrap_type_static_ptr_tag) ;
+
+}
+
+}
+
+#include <Rcpp.h>
+
+RCPP_EXPOSED_ENUM_NODECL(arrow::Type::type)
+RCPP_EXPOSED_ENUM_NODECL(arrow::DateUnit)
+RCPP_EXPOSED_ENUM_NODECL(arrow::TimeUnit::type)
+RCPP_EXPOSED_ENUM_NODECL(arrow::StatusCode)
+
+namespace Rcpp{
+namespace traits{
+
+template <typename T>
+class Exporter<std::shared_ptr<T>> {
```

Comment:
Is this one of the first R packages to manage object lifetime with `shared_ptr<T>`?

Reply:
Rcpp does not generally know how to convert a `std::shared_ptr` to an R object. What this does here is very specific to this package, and not generic. 

When `const std::shared_ptr<T>& x` is received as an input, the R object that is actually passed in is the R6 wrapper around the external pointer, so it looks for it in the environment where R6 hides it. 

When a function returns a `std::shared_ptr<T>` the function simply return a new external pointer to this shared pointer, and the R side takes over and create a new R6 object. 

I should write some blog post about this I guess.